blog/2021-09-21-intro-to-isolated-vm.md

```diff
@@ -0,0 +1,163 @@
+---
+tags:
+  - temporal
+  - community
+posted_on_: 2021-09-14T00:00:00Z
+slug: intro-to-isolated-vm
+title: Introduction to isolated-vm in Node.js
+author: Valeri Karpov
+author_title: Community Member
+author_image_url: https://avatars.githubusercontent.com/u/1620265?v=4
+release_version: V1.9.1
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+<!--truncate-->
+
+## Introduction
+
+[Temporal's Node SDK](/docs/node/introduction) makes heavy use of the [isolated-vm npm package](https://www.npmjs.com/package/isolated-vm) to ensure your [Workflows are deterministic](/docs/node/determinism).
+Each Workflow runs in a separate _isolate_ that prevents your Workflow code from directly accessing any logic that may break Workflow determinism, like reading from the file system or generating a random number.
+In this blog post, I'll cover why the Node SDK uses isolated-vm, sketch out how the Node SDK uses isolated-vm, and describe what that means for how you write Workflows in Node.js.
+
+## Why Isolates?
+
+The key idea behind Temporal is that each Workflow is stored in the Temporal server as an [_event history_](/blog/temporal-tips-tricks-1/#event-history).
+Instead of storing the "current state" of each Workflow, Temporal instead stores the Workflow's initial state, all Signals the Workflow has received, and the result of all Activities the Workflow executed.
+
+If the Temporal server wants to pause the Workflow and resume it on a different Worker, the Temporal server will replay the Workflow function from the beginning using the same Signals and Activity results.
+However, replaying event histories assumes that the Workflow is fully _deterministic_.
+That means the Workflow will be in the exact same state given the same initial state and same sequence of events.
+
+There are a lot of potential sources of non-determinism.
+Some are obvious, some are more subtle.
+For example, suppose you tried to write a Workflow that recorded the time the Workflow started, waited for a while, and then returned the time the Workflow started.
+
+```ts
+import { sleep } from '@temporalio/workflow';
+
+export const sleepWorkflow = () => {
+  async execute(): Promise<Date> {
+    const start = new Date();
+    await sleep(10 * 1000);
```

Comment:
This example isn't clear enough IMO.
This workflow is deterministic except for its result.
Maybe sleep conditionally:
```ts
if (start > new Date('2000-01-01')) {
  await sleep(10 * 1000);
}
```

Maybe even nicer if you use random.